File: packages/engine-python/src/nomark_engine/classifier.py
"""Input classification (MEE Spec Section 1.1).

Ports packages/engine/src/classifier.ts.

Tier 0: Pass-through — already resolved. Confirmations, selections, JSON, exit codes.
Tier 1: Routing — match to established pattern. Skill invocations, continuations, corrections.
Tier 2: Extraction — full intent reconstruction through resolver + gate.
"""

from __future__ import annotations

import json
import re
from dataclasses import dataclass
from typing import Literal, Protocol


InputTier = Literal[0, 1, 2]


@dataclass(frozen=True)
class ClassificationResult:
    tier: InputTier
    reason: str


class ClassifierRule(Protocol):
    tier: InputTier
    reason: str

    def test(self, input: str) -> bool: ...


@dataclass(frozen=True)
class _Rule:
    tier: InputTier
    reason: str
    _test: object  # callable

    def test(self, input: str) -> bool:
        return self._test(input)  # type: ignore[operator]


def _is_confirmation(s: str) -> bool:
    return bool(re.match(
        r"^(y|yes|no|n|ok|done|skip|cancel|approve|reject|confirm|confirmed)$",
        s.strip(), re.IGNORECASE,
    ))


def _is_numeric(s: str) -> bool:
    return bool(re.match(r"^[0-9]+$", s.strip()))


def _is_json(s: str) -> bool:
    t = s.strip()
    if not (t.startswith("{") or t.startswith("[")):
        return False
    try:
        json.loads(s)
        return True
    except (json.JSONDecodeError, ValueError):
        return False


def _is_exit(s: str) -> bool:
    return bool(re.match(r"^(exit|quit|bye|stop)\s*$", s.strip(), re.IGNORECASE))


def _is_hash(s: str) -> bool:
    return bool(re.match(r"^[a-f0-9]{6,40}$", s.strip(), re.IGNORECASE))


def _is_skill(s: str) -> bool:
    return bool(re.match(r"^/\w", s.strip()))


def _is_continuation(s: str) -> bool:
    return bool(re.match(
        r"^(continue|go ahead|proceed|next|keep going|resume)\s*$",
        s.strip(), re.IGNORECASE,
    ))


def _is_correction(s: str) -> bool:
    t = s.strip()
    return bool(
        re.match(r"^(no[,.]?\s+(not that|wrong|different|the other|I meant))", t, re.IGNORECASE)
        or re.match(r"^(actually|wait|hold on|scratch that|never\s?mind)", t, re.IGNORECASE)
    )


def _is_letter_selection(s: str) -> bool:
    t = s.strip()
    return 0 < len(t) <= 3 and bool(re.match(r"^[a-z]$", t, re.IGNORECASE))


_TIER_0_RULES: list[_Rule] = [
    _Rule(0, "confirmation", _is_confirmation),
    _Rule(0, "numeric_selection", _is_numeric),
    _Rule(0, "json_data", _is_json),
    _Rule(0, "exit_signal", _is_exit),
    _Rule(0, "hash_or_id", _is_hash),
]

_TIER_1_RULES: list[_Rule] = [
    _Rule(1, "skill_invocation", _is_skill),
    _Rule(1, "continuation", _is_continuation),
    _Rule(1, "correction", _is_correction),
    _Rule(1, "letter_selection", _is_letter_selection),
]


def classify(input: str, custom_rules: list[_Rule] | None = None) -> ClassificationResult:
    """Classify input into Tier 0 (pass-through), Tier 1 (routing), or Tier 2 (extraction)."""
    trimmed = input.strip()

    if not trimmed:
        return ClassificationResult(tier=0, reason="empty_input")

    if custom_rules:
        for rule in custom_rules:
            if rule.test(trimmed):
                return ClassificationResult(tier=rule.tier, reason=rule.reason)

    for rule in _TIER_0_RULES:
        if rule.test(trimmed):
            return ClassificationResult(tier=rule.tier, reason=rule.reason)

    for rule in _TIER_1_RULES:
        if rule.test(trimmed):
            return ClassificationResult(tier=rule.tier, reason=rule.reason)

    return ClassificationResult(tier=2, reason="substantive_input")

File: packages/engine/src/classifier.ts
/**
 * Input Classification (MEE Spec Section 1.1).
 *
 * Tier 0: Pass-through — already resolved. Confirmations, selections, JSON, exit codes.
 * Tier 1: Routing — match to established pattern. Skill invocations, continuations, corrections.
 * Tier 2: Extraction — full intent reconstruction through resolver + gate.
 */

export type InputTier = 0 | 1 | 2

export type ClassificationResult = {
  tier: InputTier
  reason: string
}

export type ClassifierRule = {
  tier: InputTier
  test: (input: string) => boolean
  reason: string
}

const TIER_0_PATTERNS: ClassifierRule[] = [
  { tier: 0, test: (s) => /^(y|yes|no|n|ok|done|skip|cancel|approve|reject|confirm|confirmed)$/i.test(s.trim()), reason: 'confirmation' },
  { tier: 0, test: (s) => /^[0-9]+$/.test(s.trim()), reason: 'numeric_selection' },
  { tier: 0, test: (s) => { try { JSON.parse(s); return s.trim().startsWith('{') || s.trim().startsWith('[') } catch { return false } }, reason: 'json_data' },
  { tier: 0, test: (s) => /^(exit|quit|bye|stop)\s*$/i.test(s.trim()), reason: 'exit_signal' },
  { tier: 0, test: (s) => /^[a-f0-9]{6,40}$/i.test(s.trim()), reason: 'hash_or_id' },
]

const TIER_1_PATTERNS: ClassifierRule[] = [
  { tier: 1, test: (s) => /^\/\w/.test(s.trim()), reason: 'skill_invocation' },
  { tier: 1, test: (s) => /^(continue|go ahead|proceed|next|keep going|resume)\s*$/i.test(s.trim()), reason: 'continuation' },
  { tier: 1, test: (s) => /^(no[,.]?\s+(not that|wrong|different|the other|I meant))/i.test(s.trim()), reason: 'correction' },
  { tier: 1, test: (s) => /^(actually|wait|hold on|scratch that|never\s?mind)/i.test(s.trim()), reason: 'correction' },
  { tier: 1, test: (s) => s.trim().length > 0 && s.trim().length <= 3 && /^[a-z]$/i.test(s.trim()), reason: 'letter_selection' },
]

/**
 * Classify input into Tier 0 (pass-through), Tier 1 (routing), or Tier 2 (extraction).
 *
 * Classification determines computational resources allocated. Tier 0 and 1 inputs
 * do not consume Signal Units.
 *
 * Custom rules are evaluated first and can override defaults.
 */
export function classify(input: string, customRules: ClassifierRule[] = []): ClassificationResult {
  const trimmed = input.trim()

  if (!trimmed) {
    return { tier: 0, reason: 'empty_input' }
  }

  // Custom rules take priority
  for (const rule of customRules) {
    if (rule.test(trimmed)) {
      return { tier: rule.tier, reason: rule.reason }
    }
  }

  // Tier 0 checks
  for (const rule of TIER_0_PATTERNS) {
    if (rule.test(trimmed)) {
      return { tier: rule.tier, reason: rule.reason }
    }
  }

  // Tier 1 checks
  for (const rule of TIER_1_PATTERNS) {
    if (rule.test(trimmed)) {
      return { tier: rule.tier, reason: rule.reason }
    }
  }

  // Default: Tier 2 (substantive NL requiring full extraction)
  return { tier: 2, reason: 'substantive_input' }
}
